### PR TITLE
test: refactor with minor fixes & improvements

### PR DESCRIPTION
**Pull request by erisu**
### Motivation and Context

* Increase coverage on code branching
* Cleanup and add better tests
* Enforce 100% coverage

### Description

Contains first phase of refactoring. Will continue in other files as well as revisiting the same files to continue with code cleanup.

Updated tests:
* handler
* parser
* update
* PackageJsonParser
* ManifestJsonParser
* Api
* Create

### Testing

- `npm t`

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [x] I added automated test

### CODE REVIEW

File: bin/templates/cordova/Api.js [hunk 1]
```diff
@@ -23,16 +23,17 @@
...
-const ActionStack = CordovaCommon.ActionStack;
-const selfEvents = CordovaCommon.events;
-// const xmlHelpers = CordovaCommon.xmlHelpers;
-const PlatformJson = CordovaCommon.PlatformJson;
-const PlatformMunger = CordovaCommon.ConfigChanges.PlatformMunger;
-const PluginInfoProvider = CordovaCommon.PluginInfoProvider;
-
+const {
+    ActionStack,
+    ConfigChanges,
```
Comment (raphinesse) [inline bin/templates/cordova/Api.js]:
```suggestion
    ConfigChanges: { PlatformMunger },
```

File: bin/templates/cordova/Api.js [hunk 2]
```diff
@@ -23,16 +23,17 @@
...
+    CordovaLogger,
+    events,
```
Comment (raphinesse) [inline bin/templates/cordova/Api.js]:
```suggestion
    events: selfEvents,
```

File: bin/templates/cordova/Api.js [hunk 3]
```diff
@@ -23,16 +23,17 @@
...
+    PlatformJson,
+    PluginInfo,
+    PluginInfoProvider
+} = require('cordova-common');
+const selfEvents = events;
```
Comment (raphinesse) [inline bin/templates/cordova/Api.js]:
```suggestion
```

File: bin/templates/cordova/Api.js [hunk 4]
```diff
@@ -23,16 +23,17 @@
...
+const PlatformMunger = ConfigChanges.PlatformMunger;
```
Comment (raphinesse) [inline bin/templates/cordova/Api.js]:
```suggestion
```

File: bin/templates/cordova/Api.js [hunk 5]
```diff
@@ -94,7 +95,7 @@ class Api {
     }
 
     addPlugin (pluginInfo, installOptions) {
-        if (!pluginInfo) {
+        if (!pluginInfo || !(pluginInfo instanceof PluginInfo)) {
```
Comment (raphinesse) [inline bin/templates/cordova/Api.js]: This could bite us. `pluginInfo instanceof PluginInfo` would be `false` if we get passed an instance of `PluginInfo` from a `cordova-common` version different from our own, even if it would be API-compatible.

File: bin/templates/cordova/Api.js [hunk 6]
```diff
@@ -144,6 +145,10 @@ class Api {
     }
 
     removePlugin (plugin, uninstallOptions) {
+        if (!plugin || !(plugin instanceof PluginInfo)) {
```
Comment (raphinesse) [inline bin/templates/cordova/Api.js]: Same problem with `pluginInfo instanceof PluginInfo`

File: package.json [hunk 7]
```diff
@@ -50,6 +50,12 @@
   },
   "nyc": {
     "all": true,
+    "check-coverage": true,
+    "per-file": true,
+    "lines": 100,
+    "statements": 100,
+    "functions": 100,
+    "branches": 100,
```
Comment (raphinesse) [inline package.json]: Maybe it would be good to allow for a small percentage of uncovered items? Sometimes things are very hard to test and people might change the code for the worse just to get the 100%. Just food for thought. Might also be OK as-is.

File: tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js [hunk 8]
```diff
@@ -28,211 +30,148 @@ const cfg1 = new ConfigParser(path.join(FIXTURES, 'test-config-1.xml'));
...
-describe('Testing ManifestJsonParser.js:', () => {
-    let ManifestJsonParser;
-    let locations;
+const locations = {
+    buildRes: path.join('mock', 'build-res'),
+    www: path.join('mock', 'www'),
+    configXml: path.join('mock', 'config.xml')
+};
+
+const defaltInitManifest = {
```
Comment (raphinesse) [inline tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js]: Typo
```suggestion
const defaultInitManifest = {
```

File: tests/spec/unit/create.spec.js [hunk 9]
```diff
@@ -49,7 +49,7 @@ function createAndVerify (projectname, projectid) {
     fs.removeSync(tmpDir);
 }
 
-function createAndValidateProjectDirName (projectname, projectid) {
+function createAndValidateProjectDirName (projectname, projectid, copyNodeModules) {
```
Comment (raphinesse) [inline tests/spec/unit/create.spec.js]: I would generally prefer something like the following, which makes the meaning of the flag obvious on call sites of the function.

```suggestion
function createAndValidateProjectDirName (projectname, projectid, { copyNodeModules = false } = {}) {
```

File: tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js [hunk 10]
```diff
@@ -28,211 +30,148 @@ const cfg1 = new ConfigParser(path.join(FIXTURES, 'test-config-1.xml'));
...
     beforeEach(() => {
-        ManifestJsonParser = rewire('../../../../../../bin/templates/cordova/lib/ManifestJsonParser');
+        manifestJsonParser = new ManifestJsonParser(locations.www);
+    });
+
+    it('should have been constructed with initial values.', () => {
+        expect(manifestJsonParser).toBeDefined();
+        expect(manifestJsonParser.path).toEqual(path.join(locations.www, 'manifest.json'));
+        expect(manifestJsonParser.www).toEqual(locations.www);
+        expect(manifestJsonParser.manifest).toEqual(defaltInitManifest);
```
Comment (erisu) [inline tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js]:
```suggestion
        expect(manifestJsonParser.manifest).toEqual(defaltInitManifest);
```

File: tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js [hunk 11]
```diff
@@ -28,211 +30,148 @@ const cfg1 = new ConfigParser(path.join(FIXTURES, 'test-config-1.xml'));
...
+    });
+
+    it('should return when config xml is not defined.', () => {
+        manifestJsonParser.configure();
+        expect(manifestJsonParser.manifest).toEqual(defaltInitManifest);
```
Comment (erisu) [inline tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js]:
```suggestion
        expect(manifestJsonParser.manifest).toEqual(defaultInitManifest);
```

File: tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js [hunk 12]
```diff
@@ -28,211 +30,148 @@ const cfg1 = new ConfigParser(path.join(FIXTURES, 'test-config-1.xml'));
...
+        spyOn(fs, 'writeFileSync').and.returnValue(true);
 
-        locations = {
-            buildRes: path.join('mock', 'build-res'),
-            www: path.join('mock', 'www'),
-            configXml: path.join('mock', 'config.xml')
+        manifestJsonParser.write();
+        expect(fs.writeFileSync).toHaveBeenCalledWith(
+            jasmine.any(String),
+            JSON.stringify(defaltInitManifest, null, 2),
```
Comment (erisu) [inline tests/spec/unit/templates/cordova/lib/ManifestJsonParser.spec.js]:
```suggestion
            JSON.stringify(defaultInitManifest, null, 2),
```

File: package.json (hunk 7 above)
Comment (erisu) [reply package.json]: Personally, I would like to keep the restrictions to try and maintain a high testing standard, even if many do hate writing tests.

I can see your point of view where it can be seen as negative.

In recent experience in our other platforms, I made changes that didn't fail the test but I found later that I broke the emulator process. This is because the area was not tested before.

In other projects for example Webkit, they have a strong requirement but I don't know if they enforce at the test tooling level.

https://webkit.org/contributing-code/#regression-tests

In summary of the Webkit,
From my point of view, they seem to always want things tested, either for a feature or fix. It does bring up a case where tests not provided can be accepted BUT the user must explain why their changes does not require tests.

File: tests/spec/unit/create.spec.js [hunk 13]
```diff
@@ -59,7 +59,15 @@ function createAndValidateProjectDirName (projectname, projectid) {
     const _fs = create.__get__('fs');
     create.__set__('fs', {
         ensureDirSync: _fs.ensureDirSync,
-        existsSync: path => !(path === projectPath),
+        existsSync: path => {
+            if (path === projectPath) {
```
Comment (GedasGa) [inline tests/spec/unit/create.spec.js]: Just a nitpick, but you could join if and else if blocks together with OR operator. Or it is done this way intentionally?

File: tests/spec/unit/templates/parser.spec.js [hunk 14]
```diff
@@ -36,201 +37,150 @@ describe('Parser class', () => {
...
+
+                const cordovaProject = {
+                    root: 'mock',
+                    projectConfig: {
+                        path: path.join('mock', 'config.xml'),
+                        cdvNamespacePrefix: 'cdv'
+                    },
+                    locations: {
+                        www: path.join('mock', 'www')
+                    }
```
Comment (GedasGa) [inline tests/spec/unit/templates/parser.spec.js]: I noticed an identical mock in this file. Maybe you could export it somewhere up above and re-use it.

File: tests/spec/unit/create.spec.js (hunk 13 above)
Comment (erisu) [reply tests/spec/unit/create.spec.js]: I did it intentionally but changed it.

File: tests/spec/fixtures/testplugin-empty-jsmodule/src/electron/sample.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+  "title" : "sample"
+}
```
[the remark this answers is not in the document]
Comment (erisu) [reply tests/spec/fixtures/testplugin-empty-jsmodule/src/electron/sample.json]: Not all of them are actually needed. I went ahead and removed some of the extra files to shrink the test fixtures.
